crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+        let (join_cluster_tx, join_cluster_rx) = mpsc::channel(1);
+
+        let rocksdb_storage =
+            RocksDbStorage::create(&Configuration::pinned().metadata_store, rocks_db_options)
+                .await?;
+
+        // make sure that the storage is initialized with a storage id to be able to detect disk losses
+        let storage_id = if let Some(storage_id) = rocksdb_storage
+            .get_storage_id()
+            .map_err(|err| BuildError::InitStorage(err.to_string()))?
+        {
+            storage_id
+        } else {
+            let storage_id = random();
+            rocksdb_storage
+                .set_storage_id(storage_id)
+                .map_err(|err| BuildError::InitStorage(err.to_string()))?;
+            storage_id
+        };
+
+        debug!("Obtained storage id: {storage_id}");
+
+        // todo also store and validate that the node name has not changed to prevent assigning a different PlainNodeId
+
+        Ok(Self {
+            connection_manager: Arc::default(),
+            rocksdb_storage,
+            storage_id,
+            metadata_writer,
+            request_tx,
+            request_rx,
+            provision_tx,
+            provision_rx: Some(provision_rx),
+            join_cluster_tx,
+            join_cluster_rx,
+        })
+    }
+
+    pub(crate) fn request_sender(&self) -> RequestSender {
+        self.request_tx.clone()
+    }
+
+    pub(crate) fn provision_sender(&self) -> ProvisionSender {
+        self.provision_tx.clone()
+    }
+
+    pub(crate) fn connection_manager(
+        &self,
+    ) -> Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>> {
+        Arc::clone(&self.connection_manager)
+    }
+
+    pub(crate) fn join_cluster_handle(&self) -> JoinClusterHandle {
+        JoinClusterHandle::new(self.join_cluster_tx.clone())
+    }
+
+    pub async fn run(self) -> anyhow::Result<()> {
+        let mut shutdown = std::pin::pin!(cancellation_watcher());
+
+        tokio::select! {
+            _ = &mut shutdown => {
+                debug!("Shutting down OmniPaxosMetadataStore");
+            },
+            result = self.run_inner() => {
+                result.context("OmniPaxosMetadataStore failed")?;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn run_inner(mut self) -> anyhow::Result<Never> {
+        if let Some(metadata_writer) = self.metadata_writer.as_mut() {
+            // Try to read a persisted nodes configuration in order to learn about the addresses of our
+            // potential peers and the metadata store states.
+            if let Some(nodes_configuration) = self
+                .rocksdb_storage
+                .get_nodes_configuration()
+                .map_err(|err| BuildError::InitStorage(err.to_string()))?
+            {
+                metadata_writer
+                    .update(Arc::new(nodes_configuration))
+                    .await?
+            }
+        }
```

Comment:
Some of the information like what are the node's addresses or what is my `MetadataStoreState` are stored in the `NodesConfiguration`. Since we cannot guarantee that we can join a cluster w/o it (e.g. when restarting all nodes at the same time or when failing over during a reconfiguration where we haven't received the omnipaxos configuration yet), we persist it ourselves.